app/component/itinerary/NaviTop.js

```diff
@@ -176,46 +216,60 @@ function NaviTop({ focusToLeg, time, realTimeLegs }, { intl, config }) {
     const newLeg = realTimeLegs.find(leg => {
       return legTime(leg.start) <= time && time <= legTime(leg.end);
     });
+
     const notifs = [];
-    if (newLeg?.id !== currentLeg?.id) {
-      if (newLeg) {
-        if (!newLeg.transitLeg) {
-          // When the component is first rendered, there is no currentLeg
-          const l = currentLeg || newLeg;
-          nextLeg = realTimeLegs.find(
-            leg => legTime(leg.start) > legTime(l.start),
-          );
+    const isSame = newLeg?.id
+      ? newLeg.id === currentLeg?.id
+      : currentLeg?.mode === newLeg?.mode;
+    const l = currentLeg || newLeg;
+    if (l) {
+      nextLeg = realTimeLegs.find(leg => legTime(leg.start) > legTime(l.start));
+      if (nextLeg) {
+        const i = getScheduleInfo(nextLeg, intl);
+        if (i) {
+          const found = notifications.find(n => n.id === i.id);
+          if (!found) {
+            notifs.push(i);
+          }
         }
-        // focus to a changed leg if not tracking location
-        focusToLeg?.(newLeg);
-        if (nextLeg) {
-          notifs.push(getScheduleInfo(nextLeg, intl));
+      }
+      let currentNots = displayNotifs;
+      if (!isSame) {
+        // remove Old main notification when new leg is started.
+        currentNots = currentNots.filter(n => n.type !== 'main');
+        if (newLeg) {
+          focusToLeg?.(newLeg);
+          setCurrentLeg(newLeg);
+        }
+        const alerts = getAlerts(realTimeLegs, intl);
+        if (alerts.length > 0) {
```

Comment:
This is not correct. Alerts can arise anytime during the trip, not only when current leg changes.